keps/sig-storage/20190120-execution-hook-design.md

```diff
@@ -0,0 +1,537 @@
+
+The Create event of ExecutionHook will trigger the hook controller to run HookAction command. The status will be updated once it is available by controller. It is the caller's (user or controller) responsibility to delete the hook once the execution finishes (either succeeded or failed). Otherwise, ExecutionHook controller will garbage collect them after a certain amount of time.
+
+Here is the definition of an ExecutionHook API object:
+
+```
+// ExecutionHook is in the tenant namespace
+type ExecutionHook struct {
+        metav1.TypeMeta
+        // +optional
+        metav1.ObjectMeta
+
+        // Spec defines the behavior of a hook.
+        // +optional
+        Spec ExecutionHookSpec
+
+        // Status defines the current state of a hook.
+        // +optional
+        Status ExecutionHookStatus
+}
+```
+
+Here is the definition of the ExecutionHookSpec:
+
+```
+// HookActionName is copied to ExecutionHookSpec by the controller such as
+// the Snapshot Controller.
+type ExecutionHookSpec struct {
+        // PodSelection defines how to select pods and containers to run
+	// the executionhook. If multiple pod/containers are selected, the action will executed on them
+	// asynchronously. If execution ordering is required, caller has to implement the logic and create
+	// different hooks in order.
+	// This field is required.
+        PodSelection PodSelection
+
+        // Name of the HookAction. This is required.
+        ActionName string
+}
+
+// PodSelection contains two fields, PodContainerNamesList and PodContainerSelector,
+// where one of them must be defined so that the hook controller knows where to
+// run the hook.
+type PodSelection struct {
+        // PodContainerNamesList lists the pods/containers on which the ExecutionHook
+        // should be executed. If not specified, the ExecutionHook controller will find
+        // all pods and containers based on PodContainerSelector.
+        // If both PodContainerNamesList and PodContainerSelector are not
+        // specified, the ExecutionHook cannot be executed and it will fail.
+        // +optional
+        PodContainerNamesList []PodContainerNames
+
+        // PodContainerSelector is for hook controller to find pods and containers
+        // based on the pod label selector and container names
+        // If PodContainerNamesList is specified, this field will not be used.
+        // +optional
+        PodContainerSelector *PodContainerSelector
```

Comment:
I think this can just be called `Selector` and be a simple selector.  Container Name should be in the action.  If the action were defined in Pod, it would be under a container, so it belongs on the other side of this relationship.